diskqueue.go

```diff
@@ -395,6 +436,184 @@ func (d *diskQueue) readOne() ([]byte, error) {
 	return readBuf, nil
 }
 
+func (d *diskQueue) removeBadFile(oldestBadFileInfo os.FileInfo) error {
+	var err error
+	badFileFilePath := path.Join(d.dataPath, oldestBadFileInfo.Name())
+
+	// remove file if it exists
+	err = os.Remove(badFileFilePath)
+	if err != nil {
+		d.logf(ERROR, "DISKQUEUE(%s) failed to remove .bad file(%s) - %s", d.name, oldestBadFileInfo.Name(), err)
+		d.updateTotalDiskSpaceUsed()
+		return err
+	} else {
+		// recaclulate total bad files disk size to get most accurate info
+		d.totalDiskSpaceUsed -= oldestBadFileInfo.Size()
+		d.logf(INFO, "DISKQUEUE(%s) removed .bad file(%s) of size(%d bytes) to free up disk space", d.name, oldestBadFileInfo.Name(), oldestBadFileInfo.Size())
+	}
+
+	return nil
+}
+
+func (d *diskQueue) removeReadFile() error {
+	var err error
+
+	if d.readFileNum == d.writeFileNum {
+		d.skipToNextRWFile()
+		return nil
+	}
+
+	if d.readFile == nil {
+		curFileName := d.fileName(d.readFileNum)
+
+		d.readFile, err = os.OpenFile(curFileName, os.O_RDONLY, 0600)
+		if err != nil {
+			return err
+		}
+
+		d.reader = bufio.NewReader(d.readFile)
+	}
+
+	closeReadFile := func() {
+		d.readFile.Close()
+		d.readFile = nil
+	}
+	defer closeReadFile()
+
+	// read total messages number at the end of the file
+	_, err = d.readFile.Seek(-numFileMsgBytes, 2)
+	if err != nil {
+		return err
+	}
+
+	var totalMessages int64
+	err = binary.Read(d.reader, binary.BigEndian, &totalMessages)
+	if err != nil {
+		return err
+	}
+
+	// update depth with the remaining number of messages
+	d.depth -= totalMessages - d.readMessages
```

Comment:
If d.readMessages is not sync well. d.readMessages is small and d.depth might be negative.
Could we add a check here?
```go
	if d.depth < 0 {
		d.logf(ERROR, "DISKQUEUE(%s) d.depth = %d is negative", d.name, d.depthChan)
		d.depth = 0
	}
```

Reply:
In that case, what if we only do that subtraction if totalMessages is greater than d.readMessages and less than d.depth.
```
if totalMessages > d.readMessages && totalMessages < d.depth {
	// update depth with the remaining number of messages
	d.depth -= totalMessages - d.readMessages
}
```
This way, we still have a rough estimate of what the total depth is. In the situation where the depth is 1,000 and a file has 100 messages. I think it would be better to have a rough estimate as to what the depth is rather than set it to 0. Plus, this is what happens when we encounter a bad file or a file that may have been truncated: we do not subtract depth by anything and proceed with an estimated number.

Reply:
We have decided that there are two problems: 1) when the current depth is a negative number and 2) when totalMessages is a negative number. I think we should approach the second problem with my suggestion
and the first problem by adding your check when we create a new DiskQueue object.

However, we will hold off on making these changes until we find the real issue of the negative number.